src/hexen/p_enemy.c

```diff
@@ -528,25 +528,29 @@ boolean P_LookForPlayers(mobj_t * actor, boolean allaround)
     player_t *player;
     angle_t an;
     fixed_t dist;
+    int i = 0; // For breaking vanilla infinite loop.
 
     if (!netgame && players[0].health <= 0)
     {                           // Single player game and player is dead, look for monsters
         return (P_LookForMonsters(actor));
     }
     c = 0;
 
-    // NOTE: This behavior has been changed from the Vanilla behavior, where
-    // an infinite loop can occur if players 0-3 all quit the game. Although
-    // technically this is not what Vanilla does, fixing this is highly
-    // desirable, and having the game simply lock up is not acceptable.
-    // stop = (actor->lastlook - 1) & 3;
-    // for (;; actor->lastlook = (actor->lastlook + 1) & 3)
-
-    stop = (actor->lastlook + maxplayers - 1) % maxplayers;
-    for (;; actor->lastlook = (actor->lastlook + 1) % maxplayers)
+    stop = (actor->lastlook - 1) & 3;
+    for (;; actor->lastlook = (actor->lastlook + 1) & 3)
     {
         if (!playeringame[actor->lastlook])
+        {
+            // In vanilla, an infinite loop can occur here if there are > 4
+            // players and players 0 - 3 all quit the game. Error out instead.
+            if (i++ == 4)
```

Comment:
Is the value 4 correct here? What happens if we start with 8 players and then players 7 and 0-2 quit?

I consider increment as part of an expression a bad habit. It always requires careful thinking of pre- vs post-increment value. It should be its own statement.

Reply:
> Is the value 4 correct here? What happens if we start with 8 players and then players 7 and 0-2 quit?

Yes, `4` is correct. Due to `actor->lastlook = (actor->lastlook + 1) & 3` we know that if the loop was repeated four times, it will continue repeating forever.

> I consider increment as part of an expression a bad habit. It always requires careful thinking of pre- vs post-increment value. It should be its own statement.

The same is done in the Vanilla code just a few lines later.